Preserve integer literal formatting in type hint

curiousleo opened the pull request:
Closes https://github.com/ocaml/ocaml/issues/10766, implementing "Proposal 1".

This is a more localised implementation than the implementation proposed in #10767, which changes `Asttype.constant`. It is based on [this comment](https://github.com/ocaml/ocaml/pull/10767#issuecomment-987948616).

# User-visible changes

The goal of this PR is to preserve the original formatting of integer literals (`int32`, `int64` and `nativeint`) in the "Did you mean ...?" hint.

**Before**
```
    1 | let _ = Int64.add 0xAA_BBl 2L;;
                          ^^^^^^^^
    Error: This expression has type int32 but an expression was expected of type
             int64
      Hint: Did you mean `43707L'?
```

**After**
```
    1 | let _ = Int64.add 0xAA_BBl 2L;;
                          ^^^^^^^^
    Error: This expression has type int32 but an expression was expected of type
             int64
      Hint: Did you mean `0xAA_BBL'?
```

# API changes

In `Typecore.error`:
- `Pattern_type_clash` contains an optional `Parsetree.pattern_desc` instead of `Typetree.pattern_desc`
- `Expr_type_clash` contains an optional `Parsetree.expression_desc` instead of `Typetree.expression_desc`

In both cases, the description was and is only used to output the type hint. The difference is that the `Parsetree` descriptions contain the original formatting of the integer literal, but the `Typetree` descriptions do not. Using the `Parsetree` descriptions instead of the `Typetree` descriptions allows us to output the integer literals with their original formatting in the type hint.

curiousleo commented:
Ready for review. @gasche and @Octachron, you have the context from #10767, perhaps you would be willing to take a look?

dra27 commented on testsuite/tests/typing-core-bugs/const_int_hint.ml:
```diff
@@ -2,11 +2,11 @@
 *)
 
-let _ = Int32.(add 1 2l);;
+let _ = Int32.add 1 2l;;
```
In passing, this isn’t a necessary change, right?

curiousleo replied:
That's correct. Happy to revert if you think this change does more harm (changing this line's Git history) than good (more consistent formatting).

dra27 replied:
All module uses in this test file were consistently using a local open before 🙂

curiousleo replied:
Fair point :) I've undone these changes.

Octachron commented:
Unsurprisingly, I prefer this implementation to #10767 .

Octachron commented on Changes:
```diff
@@ -259,6 +259,9 @@ OCaml 4.14.0
 
+- #10818: Preserve integer literal formatting in type hint.
+  (Leonhard Markert, reviewed by Gabriel Scherer and Florian Angeletti)
+
```
The change entry belongs to the "working version" section rather than 4.14 .

curiousleo replied:
My understanding is that 4.13.1 is the most recent released version, so if this PR was merged soon, it would be released as part of the next version, and my guess was that this next version would be 4.14.0. That and the fact that the "Working version" section is empty led me to add the changelog entry to the "OCaml 4.14.0" section.

This is guesswork based on how I've seen other projects handle "pending" release notes.

I'm happy to make the change you suggested, but I'd like to understand it. Given that 4.13.1 is the most recent released version, which changelog entries should be added to the "OCaml 4.14.0" section and which changelog entries belong in the (currently empty) "Working version" section?

Octachron replied:
The set of new features for OCaml 4.14 have already be fixed (in early November). The 4.14 branch is only integrating bug fixes currently. (There should be soon an alpha release for 4.14.0). In other words, all non-bug fixes PRs belong to the working version section. 


Ordinarily, this would have mean that your patch would have been integrated into the development branch and be a part of the 4.15 release.

However, the ordinary release cycle has been suspended for the integration of OCaml multicore into OCaml mainline. Once this integration work is done, we will release OCaml 5.00.0 as a first multicore version of OCaml (around March/April 2022); and only then we will resume the ordinary development cycles.

Consequently, your work will only be merged after the integration of multicore OCaml. And at the earliest, it will be part of the OCaml 5.01 release.

gasche replied:
Note: I think we should explain the freeze as a note high enough in the README, so that it is visible to people coming to this repository.

gasche replied:
(This could also be explained in Changes.)

curiousleo replied:
Thanks!

+1 for explaining the situation (or adding a link to the explanation) to `Changes`.

Octachron commented on typing/typecore.ml:
```diff
@@ -377,10 +377,10 @@ let unify_pat_types_return_equated_pairs ?(refine = None) loc env ty ty' =
   ignore (unify_pat_types_return_equated_pairs ?refine loc env ty ty')
 
-let unify_pat ?refine env pat expected_ty =
+let unify_pat ?refine ?desc env pat expected_ty =
```
A documentation comment and a more descriptive name would be nice. Maybe `sdesc_for_error`?

curiousleo replied:
Good idea. I added a comment and changed `desc` to `sdesc_for_hint`, that feels even more descriptive.

curiousleo commented:
Thanks for the suggestions for improvement! All comments addressed, ready for another look.

Octachron commented:
The current state looks good to me, I am adding this PR to the post-freeze milestone.

curiousleo commented:
> The current state looks good to me, I am adding this PR to the post-freeze milestone.

Awesome! I'll rebase on top of `trunk` and resolve the conflicts.

Octachron commented:
Note that you may want to wait before resolving conflicts, there is still many weeks left before the end of the  freeze and Changes is quite conflict-prone.

curiousleo commented:
> Note that you may want to wait before resolving conflicts, there is still many weeks left before the end of the freeze and Changes is quite conflict-prone.

Cool, good to know. I only saw your message after rebasing, but I'll keep in mind that conflicts in `Changes` are to be expected.

Will I be notified when I should rebase to get the PR ready for merging? (This is my first contribution to the OCaml compiler.)

Octachron commented:
Yes, I will notify you when the merge windows reopen.

Octachron commented:
@curiousleo : the merge windows has reopened if you wish to rebase your PR it can be merged.

curiousleo committed:
Preserve integer literal formatting in type hint

Before:

    1 | let _ = Int64.add 0xAA_BBl 2L;;
                          ^^^^^^^^
    Error: This expression has type int32 but an expression was expected of type
             int64
      Hint: Did you mean `43707L'?

After:

    1 | let _ = Int64.add 0xAA_BBl 2L;;
                          ^^^^^^^^
    Error: This expression has type int32 but an expression was expected of type
             int64
      Hint: Did you mean `0xAA_BBL'?

curiousleo commented:
Rebase is done!

Octachron commented on typing/typecore.ml:
```diff
@@ -382,10 +382,15 @@ let unify_pat_types_return_equated_pairs ?(refine = None) loc env ty ty' =
+   [sdesc_for_hint] is used to show integer literals in their original
+   formatting in the "Did you mean ...?" hint, see
+   [report_literal_type_constraint]. *)
+let unify_pat ?refine ?sdesc_for_hint env pat expected_ty =
```
The term `actual` type is inexact. I would suggest that you only document the argument that you added.
Moreover, it might be better to document the intent rather than the precise use case:
```ocaml
(** [sdesc_for_hint] is used by error messages to report literals in their original formatting *)
```

curiousleo replied:
Thanks for the suggestion! I've adopted it wholesale.

Octachron commented on typing/typecore.ml:
```diff
@@ -2779,12 +2784,16 @@ let name_cases default lst =
+   [sdesc_for_hint] is used to show integer literals in their original
+   formatting in the "Did you mean ...?" hint, see
+   [report_literal_type_constraint]. *)
+let unify_exp ?sdesc_for_hint env exp expected_ty =
```
Same as above: there is no `actual` expression type here and it is probably better to describe the intent behind `sdesc_for_hint` rather than a specific use case.

curiousleo replied:
Done, thanks!

Octachron reviewed (approved):
The PR looks good to me: it improves error messages by doing less work overall.

curiousleo commented:
All review comments addressed.

curiousleo commented:
@Octachron thanks for taking a look! What's the next step?

gasche commented:
Octachron is on holidays today and tomorrow, I'll go ahead and merge if the CI is happy.

curiousleo commented:
> Octachron is on holidays today and tomorrow, I'll go ahead and merge if the CI is happy.

Thanks! I [fixed](https://github.com/ocaml/ocaml/compare/0996df245590e0563a28771c7a7b5ca3e02717ee..ef8431146cb028c9d09db30bd9a327ee079d920b) the line length issue that was [flagged by CI](https://github.com/ocaml/ocaml/runs/7340403216?check_suite_focus=true#step%3A6%3A9=):

```
./typing/typecore.ml:386.81: [long-line] line is over 80 columns
./typing/typecore.ml:2785.81: [long-line] line is over 80 columns
```

You may need to re-trigger CI so it runs on the new commit, sorry.

Octachron commented:
Merged, thanks for your contribution and patience!

curiousleo commented:
> Merged, thanks for your contribution and patience!

My pleasure, thanks for the thoughtful reviews!
